src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java

```diff
@@ -130,47 +133,60 @@ else if (parent != null) {
     }
 
     /**
-     * Add next sibling.
+     * Add next sibling, pushes other siblings back.
      *
-     * @param ast
-     *        DetailAST object.
+     * @param ast DetailAST object.
      */
     public void addNextSibling(DetailAST ast) {
         clearBranchTokenTypes();
         clearChildCountCache(parent);
         if (ast != null) {
             // parent is set in setNextSibling
-            final DetailAstImpl nextSibling = getNextSibling();
+            final DetailAstImpl sibling = nextSibling;
             final DetailAstImpl astImpl = (DetailAstImpl) ast;
 
-            if (nextSibling != null) {
-                astImpl.setNextSibling(nextSibling);
-                nextSibling.previousSibling = astImpl;
+            if (sibling != null) {
+                astImpl.setNextSibling(sibling);
+                sibling.previousSibling = astImpl;
             }
 
             astImpl.previousSibling = this;
             setNextSibling(astImpl);
         }
     }
 
-    @Override
-    public void addChild(AST ast) {
+    /**
+     * Adds a new child to the current AST.
+     *
+     * @param child to DetailAST to add as child
+     */
+    public void addChild(DetailAST child) {
         clearBranchTokenTypes();
         clearChildCountCache(this);
-        if (ast != null) {
-            final DetailAstImpl astImpl = (DetailAstImpl) ast;
+        if (child != null) {
```

Comment:
`if (child != null)`
would be better if 
`if (child instanceof DetailAstImpl)`, it will be false for null

Reply:
Wouldn't this have to be the instance of the interface (DetailAST) and not the class?

Reply:
on next line after this condition ast is casted to DetailAstImpl, therefore I suggested implementation class, not interface. It can be changed to DetailAST if there are no specific methods invoked.

Reply:
@strkkk in this case, if there is a hypothetical `OtherDetailAST` here, the code should fall with an error.
You suggest just skipping this `if` in such case. This can lead to hard to find errors. I would leave it as it is.

Reply:
@pbludov for other classes current version of the code will fail anyway because there is unchecked cast on next line
```
        if (child != null) {
            final DetailAstImpl astImpl = (DetailAstImpl) child;
```

what I am suggesting is following

```
        if (child instanceof DetailAstImpl) {
            final DetailAstImpl astImpl = (DetailAstImpl) child;
```
this will include null check and also make cast safe. Same can be done for other similar code

Reply:
That's what I'm talking about. The code should fail if `child != null` and `child` are not an instance of `DetailAstImpl`.
You suggest to silently skip the branch if `child != null` and `child` are not an instance of `DetailAstImpl`.

Reply:
There can be else block to fail with some meaningful error. But anyway, it is not something critical